### PR TITLE
Add --samples to hirm binary

### PR DESCRIPTION
**Pull request by ThomasColthurst**
This PR is not ready to check in yet, because actually using --samples is way too slow in practice.  For example, the integration test in this PR

/bazel-bin/hirm --seed=1 --iters=5 assets/animals.unary --samples=5

currently hangs for over ten minutes.

However, I believe that is a problem in the current implementation of HIRM::sample_and_incorporate rather than anything this PR adds.

If/when this PR is submitted, it will address https://github.com/probcomp/hierarchical-irm/issues/178

### CODE REVIEW

**Comment by ThomasColthurst** [comment]
Updated with code to 
(1) Populate the domain_crps with the already observed entity id's before sampling and
(2) Unincorporate any entities that end up not being used because r_contains_items is true.

Even with these updates, though, the code still doesn't generate samples in a reasonable amount of time.  The reason appears to be somesort of mismatch between the entity id's observed by the irm->domains->crp.assignments and the entity id's observed by get_relation(r)->get_data().  For example, when running

./bazel-bin/hirm --seed=1 --iters=5 assets/animals.unary --samples=5 

I see only entity id's 0 through 49 being added to domain animal, and that is all irm->domains->crp.assignments sees as well, but get_relation(r)->get_data() contains entity id 50.

File: cxx/hirm.cc [hunk 1]
```diff
@@ -332,37 +331,67 @@ void HIRM::sample_and_incorporate_relation(std::mt19937* prng,
...
+  T_encoded_observations obs;
   std::map<std::string, CRP> domain_crps;
+  // Initialize the domain_crps from all the data across all the IRM's.
+  // If sample_and_incorporate ends up being used a lot more, we should
+  // move the domain_crps into the HIRM and maintain them when incorporating
+  // rather than recalculating them here.
+  for (const auto& [unused_cluster_id, irm] : irms) {
+    for (const auto& [domain_name, domain] : irm->domains) {
+      for (const auto& [item, table] : domain->crp.assignments) {
+        domain_crps[domain_name].incorporate(item, table);
```
Comment (emilyfertig) [inline cxx/hirm.cc:350]: Recall that in `domain->crp.assignments`, `item` is the entity ID and `table` is the cluster to which the entity belongs. In `domain_crps`, the "tables" are the entity IDs and the customers/"items" are just values from a counter of observations of that entity. So this should be:

```
int crp_item = domain_crps[domain_name].assignments.size();
domain_crps[domain_name].incorporate(crp_item, item);
```
similar to lines 370-371.

**Review by emilyfertig — CHANGES_REQUESTED** [review]

File: cxx/clean_relation.hh [hunk 2]
```diff
@@ -88,6 +88,8 @@ class CleanRelation : public Relation<T> {
     assert(!data.contains(items));
     for (int i = 0; i < std::ssize(domains); ++i) {
       domains[i]->incorporate(prng, items[i]);
+      printf("DEBUG: in clean relation %s, adding item %d to domain %s\n",
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply cxx/clean_relation.hh]: Done.

File: cxx/hirm.cc [hunk 3]
```diff
@@ -332,37 +331,67 @@ void HIRM::sample_and_incorporate_relation(std::mt19937* prng,
...
-          sample_and_incorporate_relation(prng, r, entities);
+          std::string value = sample_and_incorporate_relation(prng, r, entities);
           ++num_samples;
+          obs[r].push_back(std::make_tuple(entities, value));
+        } else {
+          // Un-incorporate the entities from the domain_crps.
+          for (size_t i = 0; i < entities.size(); ++i) {
+            domain_crps[domain_and_crps[i].first].unincorporate(
+                domain_and_crps[i].second);
+            printf("Debug: removed %s: %d \n", domain_and_crps[i].first.c_str(), domain_and_crps[i].second);
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply cxx/hirm.cc]: Done.

File: cxx/hirm.cc (hunk 1 above)
Comment (ThomasColthurst) [reply cxx/hirm.cc]: That fixed it, thanks.

File: cxx/clean_relation.hh [hunk 4]
```diff
@@ -107,11 +109,12 @@ class CleanRelation : public Relation<T> {
     clusters.at(z)->incorporate(value);
   }
 
-  void incorporate_sample(std::mt19937* prng, const T_items& items) {
+  ValueType incorporate_sample(std::mt19937* prng, const T_items& items) {
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply cxx/clean_relation.hh]: Done.

**Comment by ThomasColthurst** [comment]
Besides the fix to sampling, this commit also updates write_observations to be able to handle entity id's that were created during sampling.  I went with the boring "new_[DomainName]_[EntityId]" naming scheme for these; let me know if you have a better idea.

**Review by emilyfertig — APPROVED** [review]
Thanks!

**Pull request merged**